ENH: EADtype._find_compatible_dtype

jbrockmendel opened the pull request:
- [x] closes #24246
- [x] closes #45349
- [x] closes #32346
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Tests copied from #52833, so this may close #52235.

Needs a lot of cleanup, in particular docs.  The ArrowDtype._find_compatible_dtype implementation is quite kludgy, would appreciate help from someone more knowledgeable.

I think `_find_compatible_dtype` is the "correct" long-term non-kludge solution for setitem-with-expansion.

jbrockmendel committed:
ENH: EADtype._maybe_promote

mroeschke commented on pandas/core/arrays/arrow/dtype.py:
```diff
@@ -321,3 +326,27 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
         arr = array.cast(self.pyarrow_dtype, safe=True)
         return array_class(arr)
+
+    def _maybe_promote(self, item: Any) -> tuple[DtypeObj, Any]:
```
Shouldn't this be implemented this on `ExtensionDtype` too?

jbrockmendel replied:
it is https://github.com/pandas-dev/pandas/pull/53106/files#diff-2e345cb935fc3bae33ad469d4187e9607f679f19e30afb6d764f7c7c4f32639dR394

mroeschke replied:
Ah gotcha. I didn't notice that. Would be good for the base function to have a docstring (as you mentioned in the OP) so it's clear what this method is supposed to do

mroeschke commented on pandas/core/arrays/arrow/dtype.py:
```diff
@@ -321,3 +326,27 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
+
+    def _maybe_promote(self, item: Any) -> tuple[DtypeObj, Any]:
+        if isinstance(item, pa.Scalar):
+            if not item.is_valid:
```
Shouldn't NA always be able to be inserted into ArrowExtensionArray?

jbrockmendel replied:
im not sure.  pyarrow nulls are typed, so we could plausibly want to disallow e.g. `<pyarrow.TimestampScalar: None>` in a pyarrow integer dtype

jbrockmendel committed:
simplify check

jbrockmendel committed:
docstring

mroeschke commented on pandas/core/indexing.py:
```diff
@@ -2091,26 +2086,14 @@ def _setitem_with_indexer_missing(self, indexer, value):
-            elif isna(value):
-                new_dtype = None
+            new_dtype = None
+            if is_list_like(value):
```
Note for ArrowDtype with `pa.list_` type, we would want to treat `value` like a scalar e.g

```
ser = pd.Series([[1, 1]], dtype=pd.ArrowDtype(pa.list_(pa.int64())))
ser[0] = [1, 2]
```

mroeschke commented on pandas/core/arrays/arrow/dtype.py:
```diff
@@ -321,3 +326,27 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
+            item = item.as_py()
+
+        elif item is None or item is libmissing.NA:
+            # TODO: np.nan? use is_valid_na_for_dtype
```
Since pyarrow supports nan vs NA, possibly we want to allow nan if `pa.types.is_floating(self.pyarrow_dtype)`

jbrockmendel replied:
what to do here depends on making a decision about when/how to distinguish between np.nan and pd.NA (which i hope to finally nail down at the sprint).  doing this The Right Way would involve something like implementing EA._is_valid_na_for_dtype

mroeschke commented on pandas/core/arrays/arrow/dtype.py:
```diff
@@ -321,3 +326,27 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
+                # TODO: ask joris for help making these checks more robust
+                if item.type == self.pyarrow_dtype:
+                    return self, item.as_py()
+                if item.type.to_pandas_dtype() == np.int64 and self.kind == "i":
```
Why is this needed specifically?

jbrockmendel replied:
This was just to get the tests copied from #52833 passing.

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote

jbrockmendel commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -2281,3 +2302,29 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
+    def _maybe_promote(self, item: Any) -> tuple[DtypeObj, Any]:
+        if isinstance(item, pa.Scalar):
+            if not item.is_valid:
+                # TODO: ask joris for help making these checks more robust
```
@jorisvandenbossche any thoughts here?  (not time-sensitive)

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote

jbrockmendel committed:
Better name, tests

jbrockmendel committed:
Updat edoc

jbrockmendel commented:
Updated with some added tests, changed EADtype._maybe_promote->_find_compatible_dtype.

There are some TODOs mostly for pyarrow and related to nan-vs-NA, but I don't think any of them are blockers.

jbrockmendel committed:
Fix setitem-with-expansion for Decimal

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote

jbrockmendel commented:
_find_compatible_dtype can be used to implement a default implementation of _is_valid_scalar_for, which can be used to implement _from_scalars (both discussed on this week's dev call).

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote

phofl commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -2326,3 +2347,29 @@ def __from_arrow__(self, array: pa.Array | pa.ChunkedArray):
         return array_class(arr)
+
+    def _find_compatible_dtype(self, item: Any) -> tuple[DtypeObj, Any]:
+        if isinstance(item, pa.Scalar):
```
what happens if item is null? The pyarrow null

jbrockmendel replied:
IIUC pyarrow nulls are now typed.  id prefer to be strict about making these match, but dont care _that_ much.  am hoping @jorisvandenbossche will weigh in

phofl commented on pandas/tests/extension/test_arrow.py:
```diff
@@ -2754,6 +2754,29 @@ def test_describe_timedelta_data(pa_type):
+    [
+        (pa.scalar(4, type="int32"), 4, "int32[pyarrow]"),
+        (pa.scalar(4, type="int64"), 4, "int32[pyarrow]"),
+        # (pa.scalar(4.5, type="float64"), 4, "int32[pyarrow]"),
```
What happens here?

Also what happens with a int64 scalar and int32 dtype?

jbrockmendel replied:
id want to follow the same logic we do for numpy dtypes, but was punting here in expectation of doing it in a follow-up (likely involving joris expressing an opinion)

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote

jbrockmendel committed:
Merge branch 'main' into enh-maybe_promote
